pkg/controller/storageversiongc/gc_controller_test.go

```diff
@@ -0,0 +1,438 @@
+				},
+				{
+					APIServerID:       "kube-apiserver-3",
+					EncodingVersion:   "v2",
+					DecodableVersions: []string{"v2"},
+				},
+			},
+			CommonEncodingVersion: utilpointer.String("v1"),
+		},
+	}
+
+	clientset := fake.NewSimpleClientset(lease1, lease2, lease3, storageVersion)
+	setupController(clientset)
+
+	// Delete the lease object and verify that storage version status is updated
+	if err := clientset.CoordinationV1().Leases(metav1.NamespaceSystem).Delete(context.Background(), "kube-apiserver-2", metav1.DeleteOptions{}); err != nil {
+		t.Fatalf("error deleting lease object: %v", err)
+	}
+
+	// add a delay to ensure controller had a chance to reconcile
+	time.Sleep(time.Second)
+
+	storageVersion, err := clientset.InternalV1alpha1().StorageVersions().Get(context.Background(), "k8s.test.resources", metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("error getting StorageVersion: %v", err)
+	}
+
+	expectedServerStorageVersions := []apiserverinternalv1alpha1.ServerStorageVersion{
+		{
+			APIServerID:       "kube-apiserver-1",
+			EncodingVersion:   "v1",
+			DecodableVersions: []string{"v1"},
+		},
+		{
+			APIServerID:       "kube-apiserver-3",
+			EncodingVersion:   "v2",
+			DecodableVersions: []string{"v2"},
+		},
+	}
+
+	if !reflect.DeepEqual(storageVersion.Status.StorageVersions, expectedServerStorageVersions) {
+		t.Error("unexpected storage version object")
+		t.Logf("got: %+v", storageVersion)
+		t.Logf("expected: %+v", expectedServerStorageVersions)
+	}
+
+	if *storageVersion.Status.CommonEncodingVersion != "v1" {
+		t.Errorf("unexpected common encoding version")
+		t.Logf("got: %q", *storageVersion.Status.CommonEncodingVersion)
+		t.Logf("expected: %q", "v1")
+	}
+}
+
+// Test_StorageVersionContainsInvalidLeaseID validates that status.serverStorageVersions
+// only contains the holder identity from kube-apiserver Leases that exist.
+func Test_StorageVersionContainsInvalidLeaseID(t *testing.T) {
+	lease1 := &coordinationv1.Lease{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "kube-apiserver-1",
+			Namespace: metav1.NamespaceSystem,
+			Labels: map[string]string{
+				"k8s.io/component": "kube-apiserver",
+			},
+		},
+		Spec: coordinationv1.LeaseSpec{
+			HolderIdentity: utilpointer.StringPtr("kube-apiserver-1"),
+		},
+	}
+	lease2 := &coordinationv1.Lease{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "kube-apiserver-2",
+			Namespace: metav1.NamespaceSystem,
+			Labels: map[string]string{
+				"k8s.io/component": "kube-apiserver",
+			},
+		},
+		Spec: coordinationv1.LeaseSpec{
+			HolderIdentity: utilpointer.StringPtr("kube-apiserver-2"),
+		},
+	}
+	lease3 := &coordinationv1.Lease{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "kube-apiserver-3",
+			Namespace: metav1.NamespaceSystem,
+			Labels: map[string]string{
+				"k8s.io/component": "kube-apiserver",
+			},
+		},
+		Spec: coordinationv1.LeaseSpec{
+			HolderIdentity: utilpointer.StringPtr("kube-apiserver-3"),
+		},
+	}
```

Comment:
`lease1`,`lease2` and `lease3` are unchanged and not mutated in all the tests. Could we just initialize these variables once in this test file and reuse for all the tests?

Reply:
I was avoiding shared variables because they are passed into the fake client and there might be unintended behaviors across tests. 

But I agree we can have a function to construct these since only the name and holder identity changes

Reply:
updated with constructor function